File: python/euler/003.py
from math import sqrt

def prime_factor(num, factor=2, factors=[]):
    if is_prime(num):
        factors.append(num);
        return factors
    elif (num % factor == 0 and is_prime(factor)):
        num = int(num / factor)
        factors.append(factor)
        return prime_factor(num, factor, factors)
    else:
        factor += 1
        return prime_factor(num, factor, factors)

def is_prime(num):
    if num == 2: return True
    if num %  2 == 0: return False
    for i in range(3, int(sqrt(num) + 1), 2):
        if num % i == 0: return False
    return True

if __name__ == "__main__":
    A = prime_factor(27023)
    print(f"prime_factor() = {A}")

File: julia/euler/003.jl
function prime_factor(num::Int, factor::Int, factors::Array{Int, 1})::Array{Int, 1}
  if is_prime(num)
    append!(factors, num)
    return factors
  elseif num % factor == 0 && is_prime(factor)
    num = convert(Int, num/factor)
    append!(factors, factor)
    return prime_factor(num, factor, factors)
  else
    factor += 1
    return prime_factor(num, factor, factors)
  end
end

prime_factor(num) = prime_factor(num, 2, [1])

function is_prime(num::Int)::Bool
  if num == 2 return true end 
  if num % 2 == 0 return false end
  for i in range(3, convert(Int, round(sqrt(num))); step=2)
    if num % i == 0 return false end
  end
  return true
end


